MyOnboarding/src/MyOnboardingApp.Services/Services/UpdateItemService.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Threading.Tasks;
+using MyOnboardingApp.Contracts.Generators;
+using MyOnboardingApp.Contracts.Models;
+using MyOnboardingApp.Contracts.Repository;
+using MyOnboardingApp.Contracts.Services;
+using MyOnboardingApp.Contracts.Validation;
+
+namespace MyOnboardingApp.Services.Services
+{
+    internal class UpdateItemService : AbstractUpsertService, IUpdateItemService
+    {
+        private readonly ITodoListRepository _repository;
+        private readonly IDateTimeGenerator _dateTimeGenerator;
+
+
+        public UpdateItemService(ITodoListRepository repository, IDateTimeGenerator dateTimeGenerator, IInvariantValidator<TodoListItem> validator)
+            : base(validator)
+        {
+            _dateTimeGenerator = dateTimeGenerator;
+            _repository = repository;
+        }
+
+
+        public async Task<IItemWithErrors<TodoListItem>> EditItemAsync(TodoListItem replacingItem, IResolvedItem<TodoListItem> existingItem)
+        {
+            if (replacingItem == null)
+            {
+                throw new ArgumentNullException(nameof(replacingItem), "Item to complete must not be null.");
+            }
+
+            if (existingItem == null || !existingItem.WasOperationSuccessful)
+            {
+                throw new ArgumentException("Existing item does not actually exist.", nameof(existingItem));
+            }
+
+            var initData = GetGeneratedData(existingItem.Item);
+
+            return await TryCompleteAndStoreItemAsync(replacingItem, initData);
+        }
+
+
+        protected override async Task StoreToDatabase(TodoListItem completedItem)
+            => await _repository.ReplaceItemAsync(completedItem);
+
+
+        protected virtual ItemInitializingData GetGeneratedData(TodoListItem item)
+            => new ItemInitializingData(item.Id, item.CreationTime, _dateTimeGenerator.GetCurrentDateTime());
```

Comment:
I'm still kind of torn between calling method within other method call and bodyless function…
Please consider wrapping here:
```csharp
            => new ItemInitializingData(
                item.Id,
                item.CreationTime,
                _dateTimeGenerator.GetCurrentDateTime());
```

BTW: here is a screenshot from VS2019: 
![image](https://user-images.githubusercontent.com/15132110/55242282-33f9c380-523d-11e9-8e33-67846b3b6427.png)
You can see the parameter name there, I'm kind of undecided whether use this IDE-specific feature or prefer the original C# solution of explicit parameter naming if you decided to go with wrapping, name the parameter explicitly, please:
```csharp
            => new ItemInitializingData(
                item.Id,
                item.CreationTime,
                lastModificationTime: _dateTimeGenerator.GetCurrentDateTime());
```